backtrace-library/src/main/cpp/client-side-unwinding.cpp

```diff
@@ -0,0 +1,387 @@
+    static const char *buffer = static_cast<char *>(buffer_create());
+
+    bun_buffer_init(&buf, (char*) buffer, BUFFER_SIZE);
+
+    bun_initialized.store(bun_handle_init(&handle, BUN_BACKEND_LIBUNWINDSTACK));
+
+    if (!bun_initialized) {
+        __android_log_print(ANDROID_LOG_ERROR, "Backtrace-Android", "Local client side unwinding initialization failed");
+        return false;
+    }
+
+    return true;
+}
+
+bool InitializeRemoteClientSideUnwinding(JNIEnv* env, const char* path_cstr) {
+    struct bcd_config cf;
+    bcd_error_t e;
+
+    // Initialize a shared memory region.
+    static const char *buffer = static_cast<char *>(buffer_create());
+
+    // Initialize the BCD configuration file.
+    if (bcd_config_init(&cf, &e) == -1) {
+        __android_log_print(ANDROID_LOG_ERROR, "Backtrace-Android",
+                            "bcd_config_init failed, cannot start client side unwinding, error message %s, error code %d",
+                            e.message, e.errnum);
+        return false;
+    }
+
+    // Request handler to be called when processing errors by BCD worker.
+    cf.request_handler = request_handler;
+
+    // Set a function to be called by the child for setting permissions.
+    cf.monitor_init = monitor_init;
+
+    // Create the fully resolved path to the bcd socket file
+    const char* file_name = "/bcd.socket";
+    char* full_path = (char*) malloc((strlen(path_cstr) + strlen(file_name) + 1) * sizeof(char));
+    strcpy(full_path, path_cstr);
+    strcat(full_path, file_name);
+
+    cf.ipc.us.path = full_path;
+
+    if (bcd_init(&cf, &e) == -1) {
+        __android_log_print(ANDROID_LOG_ERROR, "Backtrace-Android",
+                            "bcd_init failed, cannot start client side unwinding, error message %s, error code %d",
+                            e.message, e.errnum);
+        return false;
+    }
+
+    // Initialize the BCD handler.
+    if (bcd_attach(&bcd, &e) == -1) {
+        __android_log_print(ANDROID_LOG_ERROR, "Backtrace-Android",
+                            "bcd_attach failed, cannot start client side unwinding, error message %s, error code %d",
+                            e.message, e.errnum);
+        return false;
+    }
+
+    buf.data = (char *)buffer;
+    buf.size = BUFFER_SIZE;
+
+    pid_t child_pid;
+    memcpy(&child_pid, buffer, sizeof(child_pid));
+
+    prctl(PR_SET_PTRACER, child_pid, 0, 0, 0);
+    prctl(PR_SET_DUMPABLE, 1);
+
+    bun_initialized = true;
+    return true;
+}
+
+bool EnableClientSideUnwinding(JNIEnv *env, const char* path, jint unwindingMode) {
+    if (initialized) {
+        __android_log_print(ANDROID_LOG_ERROR, "Backtrace-Android",
+                            "Client side unwinding needs to be enabled BEFORE crashpad initialization");
+        return false;
+    }
+
+    unwinding_mode = static_cast<UnwindingMode>((int) unwindingMode);
+
+    switch (unwinding_mode) {
+        case UnwindingMode::LOCAL:
```

Comment:
I'd add a `/* fallthrough */` comment or use the [`[[fallthrough]]`](https://en.cppreference.com/w/cpp/language/attributes/fallthrough) attribute.

Reply:
We can't assume we have C++17 so I will add the comment to make it clear!